activerecord/lib/active_record/associations/join_dependency.rb

```diff
@@ -272,7 +272,9 @@ def construct(ar_parent, parent, row, rs, seen, model_cache, aliases)
             construct(model, node, row, rs, seen, model_cache, aliases)
           else
             model = construct_model(ar_parent, node, row, model_cache, id, aliases)
-            model.readonly!
+            if node.reflection.scope_for(node.base_klass).values[:readonly]
```

Comment:
Thanks to catch it. Fixed so in https://github.com/rails/rails/pull/24099/commits/411e4597d552aa091cfe3dc76a83bdaffe02da1a.